dandiapi/api/views/version.py

```diff
@@ -153,3 +153,28 @@ def publish(self, request, **kwargs):
 
         serializer = VersionSerializer(new_version)
         return Response(serializer.data, status=status.HTTP_200_OK)
+
+    @swagger_auto_schema(
+        manual_parameters=[DANDISET_PK_PARAM, VERSION_PARAM],
+    )
+    def destroy(self, request, **kwargs):
+        """
+        Delete a version.
+
+        Deletes a version. Only published versions can be deleted, and only by
+        admin users.
+        """
+        version: Version = self.get_object()
+        if version.version == 'draft':
+            return Response(
+                'Cannot delete draft versions',
+                status=status.HTTP_403_FORBIDDEN,
```

Comment:
I am wondering if this (and the other instance below) should be `405` rather than `403`. But reading about the error codes, it seems like a toss-up. I'm ok with leaving this as is, but I wanted to mention my confusion over which is more appropriate as a discussion topic.

Reply:
I also vote 405, but the distinction is not important to me.

Reply:
from 
![image](https://user-images.githubusercontent.com/39889/129772154-56cfcf91-e082-4533-9e8b-63e7a1536088.png)
seems like `403` is a good one... the bible (https://httpwg.org/specs/rfc7231.html#status.403) confirms it:

- `403`: ... "The client SHOULD NOT automatically repeat the request with the same credentials. The client MAY repeat the request with new or different credentials. "
- `405`: also reminds that ". The origin server MUST generate an `Allow` header field in a 405 response containing a list of the target resource's currently supported methods.", which we do not want to mess here with, reconfirming that `403` is the right one to use

Reply:
Yeah I think 403 is good here. But, your quote from the bible in the 403 section is not evidence for that: in our case, a `DELETE` to certain resources will always fail, so repeating the request with new or different credentials is not indicated.

In any case, let's just stick with 403. Thanks for the discussion.

Reply:
> a `DELETE` to certain resources will always fail

to some other "resources", not to the endpoint in question, right? ;)

Reply:
Not sure what you mean, but in this case, doing a `DELETE` to any draft version of any dandiset will always result in failure, regardless of authentication or authorization level. The draft version is a `Version` resource, so that's what I meant in my last comment.

Reply:
ah, I thought that by "resources" [*] you meant some other endpoints (e.g. for a top level `/dandiset/{id}`) not the versions... 

[*] as the term used in in the [RFC, and identified by URL](https://httpwg.org/specs/rfc7231.html#resources) so means exactly what you had in mind -- "draft version of a dandiset"

indeed ! Now you convinced a blind me me in the opposite, that `405` should be the right one to use **here** (and keep 403 for admin-vs-mortal distinction), with `Allow` then listing `GET` and `PUT`? (is there programmatic way to get all known to this endpoint and just remove DELETE?)

Reply:
> ah, I thought that by "resources" [*] you meant some other endpoints (e.g. for a top level `/dandiset/{id}`) not the versions...
> 
> [*] as the term used in in the [RFC, and identified by URL](https://httpwg.org/specs/rfc7231.html#resources) so means exactly what you had in mind -- "draft version of a dandiset"
> 
> indeed ! Now you convinced a blind me me in the opposite, that `405` should be the right one to use **here** (and keep 403 for admin-vs-mortal distinction), with `Allow` then listing `GET` and `PUT`?

The only trouble I can think of is that there is in fact one use case for `DELETE`ing a draft version--when we eventually get to "archive" or "protection" semantics for dandiset, we will indeed want to remove the draft version from consideration. So maybe 403 is, in fact, the right choice here.

> (is there programmatic way to get all known to this endpoint and just remove DELETE?)

I don't think so--the allowed methods list is specific knowledge about the semantics of a given endpoint. In this case, if we wanted to use 405, we would have to edit the headers of the response object and fill in the allowed method names.

But I think we should stick with 403 for now. 405 makes more sense when there are no conceivable semantics for a specific method, which I don't believe is the case here.

Reply:
I am fine either way.
If we go with 403 - anything else to do to get it merged?

Reply:
Just Daniel's approval, which I see he just granted!